spec.emu

```diff
@@ -4,20 +4,2216 @@
+        1. Let _privateEnv_ be *null*.
+        1. For each element _d_ of _lexDeclarations_, do
+          1. NOTE: Lexically declared names are only instantiated here but not initialized.
+          1. For each element _dn_ of the BoundNames of _d_, do
+            1. If IsConstantDeclaration of _d_ is *true*, then
+              1. Perform ? <emu-meta effects="user-code">_env_.CreateImmutableBinding</emu-meta>(_dn_, *true*).
+            1. Else,
+              1. Perform ? <emu-meta effects="user-code">_env_.CreateMutableBinding</emu-meta>(_dn_, *false*).
+            1. <ins>If _d_ is a |ModuleDeclaration|, then</ins>
+              1. <ins>Append _d_ to _modulesToInitialize_.</ins>
+        1. <ins>For each Parse Node _m_ of _modulesToInitialize_, do</ins>
+            1. <ins>Let _mn_ be the sole element of the BoundNames of _m_.</ins>
+            1. <ins>Let _mo_ be InstantiateModuleDeclaration of _m_.</ins>
+            1. <ins>Perform ! _env_.InitializeBinding(_mn_, _mo_).</ins>
+        1. For each Parse Node _f_ of _functionsToInitialize_, do
+          1. Let _fn_ be the sole element of the BoundNames of _f_.
+          1. Let _fo_ be InstantiateFunctionObject of _f_ with arguments _env_ and _privateEnv_.
+          1. Perform ? <emu-meta effects="user-code">_env_.CreateGlobalFunctionBinding</emu-meta>(_fn_, _fo_, *false*).
+        1. For each String _vn_ of _declaredVarNames_, do
+          1. Perform ? <emu-meta effects="user-code">_env_.CreateGlobalVarBinding</emu-meta>(_vn_, *false*).
+        1. Return ~unused~.
+      </emu-alg>
+      <emu-note>
+        <p>Early errors specified in <emu-xref href="#sec-scripts-static-semantics-early-errors"></emu-xref> prevent name conflicts between function/var declarations and let/const/class<ins>/module</ins> declarations as well as redeclaration of let/const/class<ins>/module</ins> bindings for declaration contained within a single |Script|. However, such conflicts and redeclarations that span more than one |Script| are detected as runtime errors during GlobalDeclarationInstantiation. If any such errors are detected, no bindings are instantiated for the script. However, if the global object is defined using Proxy exotic objects then the runtime tests for conflicting declarations may be unreliable resulting in an abrupt completion and some global declarations not being instantiated. If this occurs, the code for the |Script| is not evaluated.</p>
+        <p>Unlike explicit var or function declarations, properties that are directly created on the global object result in global bindings that may be shadowed by let/const/class<ins>/module</ins> declarations.</p>
+      </emu-note>
+    </emu-clause>
+  </emu-clause>
+
+  <emu-clause id="sec-modules">
+    <h1>Modules</h1>
+    <h2>Syntax</h2>
+    <emu-grammar type="definition">
+      Module :
+        ModuleBody?
+
+      ModuleBody :
+        ModuleItemList
+
+      ModuleItemList :
+        ModuleItem
+        ModuleItemList ModuleItem
```

Comment:
Yes. If there is a good reason to allow them in nested blocks then we can do it (however, it considerably complicates the static semantics) , but since they can always be manually moved to the top-level scope it doesn't seem like a bad restriction.

Reply:
Huh, I definitely imagined that they'd be permitted nested anywhere; it's just that they'd only have the ability to be statically imported if they were at the top level. Nesting might be useful because all sorts of code transformations end up making certain top-level code nested; we should insert restrictions only where needed semantically IMO.

Nested module declarations might be less useful than top-level ones, but a natural place where you might expect them to "just work" (even though it's not really a significant use of the feature) is to declare a module and then use it as a variable (as a module block), in some nested code. For this case, one could even imagine a bundler placing a number of module fragments in the same inner nested scope, so that the module block passed to the worker closes over them.